src/Framework/Analyzer/src/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs

```diff
@@ -0,0 +1,239 @@
+                            invocation));
+                }
+
+                // var builder = WebApplication.CreateBuilder();
+                // builder.WebHost.UseStartup<Startup>();
+                if (IsDisallowedMethod(
+                        operationAnalysisContext,
+                        invocation,
+                        targetMethod,
+                        wellKnownTypes.ConfigureWebHostBuilder,
+                        "UseStartup",
+                        userStartupTypes))
+                {
+                    operationAnalysisContext.ReportDiagnostic(
+                        CreateDiagnostic(
+                            DiagnosticDescriptors.DoNotUseUseStartupWithConfigureWebHostBuilder,
+                            invocation));
+                }
+
+                static Diagnostic CreateDiagnostic(DiagnosticDescriptor descriptor, IInvocationOperation operation)
+                {
+                    // Take the location for the whole invocation operation as a starting point.
+                    var location = operation.Syntax.GetLocation();
+
+                    // As we're analyzing an extension method that might be chained off a number of
+                    // properties, we need the location to be where the invocation of the targeted
+                    // extension method is, not the beginning of the line where the chain begins.
+                    // So in the example `foo.bar.Baz(x => {})` we want the span to be for `Baz(x => {})`.
+                    // Otherwise the location can contain other unrelated bits of an invocation chain.
+                    // Take for example the below block of C#.
+                    //
+                    // builder.Host
+                    //   .ConfigureWebHost(webHostBuilder => { })
+                    //   .ConfigureSomethingElse()
+                    //   .ConfigureYetAnotherThing(x => x());
+                    //
+                    // If we did not just select the method name, the location would end up including
+                    // the start of the chain and the leading trivia before the method invocation:
+                    //
+                    // builder.Host
+                    //   .ConfigureWebHost(webHostBuilder => { })
+                    //
+                    // IdentifierNameSyntax finds non-generic methods (e.g. `Foo()`), whereas
+                    // GenericNameSyntax finds generic methods (e.g. `Foo<T>()`).
+                    var methodName = operation.Syntax
+                        .DescendantNodes()
+                        .OfType<SimpleNameSyntax>()
+                        .Where(node => node is IdentifierNameSyntax || node is GenericNameSyntax)
+                        .Where(node => string.Equals(node.Identifier.Value as string, operation.TargetMethod.Name, StringComparison.Ordinal))
+                        .FirstOrDefault();
```

Comment:
Happy to refactor this if there's a better way I didn't work out for myself.

I assumed that using LINQ here was OK as we're already on the path where a diagnostic is going to be fired.